### PR TITLE
Updated Readme for --server-connect-attribute option

### PR DESCRIPTION
**Pull request by NimishaS**
w.r.t https://github.com/chef/knife-ec2/issues/379

### CODE REVIEW

**Comment by adamedx** [comment]
@btm, can you take a look?

File: lib/chef/knife/ec2_server_create.rb [hunk 1]
```diff
@@ -1108,15 +1108,22 @@ def wait_for_direct_sshd(hostname, ssh_port)
...
-                                else
-                                  server.dns_name || server.public_ip_address
-                                end
-                              end
+        connect_attribute = ""
+        if config[:server_connect_attribute]
+          @ssh_connect_host ||= server.send(config[:server_connect_attribute])
+          connect_attribute = config[:server_connect_attribute]
+        else
+          if vpc_mode? && !config[:associate_public_ip]
```
Comment (NimishaS) [inline lib/chef/knife/ec2_server_create.rb]: @btm , we'll fix https://github.com/chef/knife-ec2/issues/344 in a separate PR

**Comment by NimishaS** [comment]
@btm, fixed the review comments. Please check.

File: README.md [hunk 2]
```diff
@@ -161,6 +161,11 @@ knife ec2 server create -I ami-d0f89fb9 --ssh-key your-public-key-id -f m1.mediu
 # A Windows instance via the WinRM protocol -- --ssh-key is still required due to EC2 API operations that need it to grant access to the Windows instance
 # `--spot-price` option lets you specify the spot pricing
 knife ec2 server create -I ami-173d747e -G windows -f m1.medium --user-data ~/your-user-data-file -x '.\a_local_user' -P 'yourpassword' --ssh-key your-public-key-id --spot-price price-in-USD
+
+# Pass --server-connect-attribute if the machine is in VPC mode
+# Possible values of --server-connect-attribute: private_dns_name, private_ip_address, public_dns_name, public_ip_address
+# If --server-connect-attribute is not specified, then EIP is used as External IP
```
Comment (btm) [inline README.md]: EIP means Elastic IP (which is a public IP). Not all instances have an Elastic IP assigned to them. So this isn't true.

I would say "If --server-connect-attribute is not specified, knife attempts to determine if connecting to the instance's public or private IP is most appropriate based on other settings."

File: README.md [hunk 3]
```diff
@@ -161,6 +161,11 @@ knife ec2 server create -I ami-d0f89fb9 --ssh-key your-public-key-id -f m1.mediu
...
+knife ec2 server create -I ami-173d747e --ssh-key your-public-key-id --ssh-user ubuntu --bootstrap-protocol ssh --subnet your-subnet --security-group-ids your-security-group-id --associate-public-ip --server-connect-attribute public_ip_address
```
Comment (btm) [inline README.md]: this is kind of an intimidatingly long example. I'd remove the options we don't need and shorten the ones that aren't relevant to --server-connect-attribute, e.g.

```knife ec2 server create -I ami-173d747e -x ubuntu --server-connect-attribute public_ip_address```

We can presume in this example that subnet, ssh key, and security group are relatively static for the user and set in the knife.rb.

File: README.md [hunk 4]
```diff
@@ -161,6 +161,11 @@ knife ec2 server create -I ami-d0f89fb9 --ssh-key your-public-key-id -f m1.mediu
 # A Windows instance via the WinRM protocol -- --ssh-key is still required due to EC2 API operations that need it to grant access to the Windows instance
 # `--spot-price` option lets you specify the spot pricing
 knife ec2 server create -I ami-173d747e -G windows -f m1.medium --user-data ~/your-user-data-file -x '.\a_local_user' -P 'yourpassword' --ssh-key your-public-key-id --spot-price price-in-USD
+
+# Pass --server-connect-attribute if the machine is in VPC mode
```
Comment (btm) [inline README.md]: I don't think this is limited to VPC, although that's where we're going to pretty much always use it.

```Pass --server-connect-attribute to specify the instance attribute that we will try to connect to via ssh/winrm```

File: lib/chef/knife/ec2_server_create.rb [hunk 5]
```diff
@@ -1108,15 +1108,21 @@ def wait_for_direct_sshd(hostname, ssh_port)
...
+            connect_attribute = "private_ip_address"
+          else
+            connect_attribute = server.dns_name ? "dns_name" : "public_ip_address"
+          end
+        end
+
+        @ssh_connect_host ||= server.send(connect_attribute)
```
Comment (btm) [inline lib/chef/knife/ec2_server_create.rb]: If you're not going to wrap the conditional logic above based on if this instance variable is set, it doesn't need to be an instance variable because the only place it is used is a few lines down. The purpose of wrapping the logic is so that logic is only parsed once per run.

File: lib/chef/knife/ec2_server_create.rb [hunk 6]
```diff
@@ -1108,15 +1108,21 @@ def wait_for_direct_sshd(hostname, ssh_port)
...
+
+        puts
+        puts "SSH Target Address: #{connect_attribute}(#{@ssh_connect_host})"
```
Comment (btm) [inline lib/chef/knife/ec2_server_create.rb]: I'd still like these to be switched.

```
puts "\nSSH Target Address: #{@ssh_connect_host}(#{connect_attribute})"
```

File: lib/chef/knife/ec2_server_create.rb [hunk 7]
```diff
@@ -1108,15 +1108,21 @@ def wait_for_direct_sshd(hostname, ssh_port)
...
-                                if vpc_mode? && !config[:associate_public_ip]
-                                  server.private_ip_address
-                                else
-                                  server.dns_name || server.public_ip_address
-                                end
-                              end
+        if config[:server_connect_attribute]
+          connect_attribute = config[:server_connect_attribute]
+        else
+          if vpc_mode? && !config[:associate_public_ip]
```
Comment (btm) [inline lib/chef/knife/ec2_server_create.rb]:
```
    if vpc_mode? && !(config[:associate_public_ip] || config[:associate_eip])
```

Comment (btm) [inline lib/chef/knife/ec2_server_create.rb]: this can be a separate PR, and needs a test whenever we add it.

File: lib/chef/knife/ec2_server_create.rb (hunk 5 above)
Comment (NimishaS) [inline lib/chef/knife/ec2_server_create.rb]: My bad, fixing it.

**Comment by NimishaS** [comment]
@btm , please verify again

**Comment by btm** [comment]
:+1:

**Comment by gbish** [comment]
Hi,

I've got some news for you, please read it, this really matters a lot! Please read here <http://thepypole.bostonirishbrawler.com/oigiztvz>

Sent from my iPhone, hebgb@blueyonder.co.uk

File: lib/chef/knife/ec2_server_create.rb (hunk 1 above)
Comment (gbish) [inline lib/chef/knife/ec2_server_create.rb]: Hi,

Have you already seen that great stuff I've just found on the web? You should definitely take a look <http://temojeje.stlmortgage.com/oigrkhq>

hebgb@blueyonder.co.uk

**Comment by gbish** [comment]
Hi,

I've never seen something like that before, it's fantastic! You should take a look  <http://syntiphaba.stephenhabgood.com/oixzdlwn>

Hope this helps, hebgb@blueyonder.co.uk

File: lib/chef/knife/ec2_server_create.rb (hunk 1 above)
Comment (gbish) [inline lib/chef/knife/ec2_server_create.rb]: Hi,

I've never seen something like that before, it's fantastic! You should take a look  <http://nonaguquo.customercareofamerica.com/oibxxa>

All best, hebgb@blueyonder.co.uk
